### PR TITLE
Improve server support in Core Business

### PR DESCRIPTION
**Pull request by defagos**
# Description

This PR improves server support in Core Business. It also enhances server selection in the demo to support SAM as well as simulated locations in CH or abroad.

# Changes made

- Make it possible to use any server conforming to the media composition data format and image resizing. 
- Remove `.mmf` from Core Business server list (development tool). This is a breaking API change.
- Add support for SAM (prod, stage, test).
- Add support for simulated CH and WW locations for IL services (helpful to test geoblocking behavior).

# Checklist

- [x] APIs have been properly documented (if relevant).
- [x] The documentation has been updated (if relevant).
- [x] New unit tests have been written (if relevant).
- [x] The demo has been updated (if relevant).

### CODE REVIEW

File: Demo/Sources/Model/ServerSetting.swift [hunk 1]
```diff
@@ -12,6 +12,19 @@ enum ServerSetting: Int, CaseIterable {
...
+    case stageCH
+    case testCH
+
+    case productionWW
+    case stageWW
+    case testWW
+
+    case samProduction
+    case samStage
+    case samTest
```
Comment (waliid) [inline Demo/Sources/Model/ServerSetting.swift:26]: As we prefix these cases with `sam`, should we also prefix the original cases with `il`?

File: Sources/CoreBusiness/DataProvider/Server.swift [hunk 2]
```diff
@@ -7,29 +7,102 @@
...
-        case .stage:
-            return URL(string: "https://il-stage.srgssr.ch")!
-        case .test:
-            return URL(string: "https://il-test.srgssr.ch")!
-        case .mmf:
-            return URL(string: "https://play-mmf.herokuapp.com")!
+    public static let test = custom(baseUrl: URL(string: "https://il-test.srgssr.ch")!)
+
+    private let requestBuilder: (String) -> URLRequest
+    private let resizedImageUrlBuilder: (URL, ImageWidth) -> URL
```
Comment (waliid) [inline Sources/CoreBusiness/DataProvider/Server.swift:27]: These two closures seem to be used many times, so my suggestion is to use type aliases.

```swift
public typealias RequestBuilder = (String) -> URLRequest
public typealias UrlBuilder = (URL, ImageWidth) -> URL
```

File: Sources/CoreBusiness/DataProvider/Server.swift [hunk 3]
```diff
@@ -7,29 +7,102 @@
...
+
+    /// Custom environment.
+    ///
+    /// - Parameters:
+    ///   - baseUrl: The base URL of the server.
+    ///   - queryItems: Additional query items to use.
+    ///
+    /// Useful for servers which can exactly pose as SRG SSR servers and deliver the same playback metadata format and
+    /// image scaling capabilities.
+    public static func custom(baseUrl: URL, queryItems: [URLQueryItem] = []) -> Self {
```
Comment (waliid) [inline Sources/CoreBusiness/DataProvider/Server.swift:61]: What do you think about renaming `custom` by `urn`?

```suggestion
    public static func urn(baseUrl: URL, queryItems: [URLQueryItem] = []) -> Self {
```

File: Sources/CoreBusiness/DataProvider/Server.swift [hunk 4]
```diff
@@ -7,29 +7,102 @@
...
+        .custom { urn in
+            standardRequest(forUrn: urn, baseUrl: baseUrl, queryItems: queryItems)
+        } resizedImageUrlBuilder: { url, width in
+            standardResizedImageUrl(url, width: width, baseUrl: baseUrl, queryItems: queryItems)
+        }
+    }
+
+    private static func standardRequest(forUrn urn: String, baseUrl: URL, queryItems: [URLQueryItem]) -> URLRequest {
```
Comment (waliid) [inline Sources/CoreBusiness/DataProvider/Server.swift:69]:
```suggestion
    private static func request(forUrn urn: String, baseUrl: URL, queryItems: [URLQueryItem]) -> URLRequest {
```

File: Sources/CoreBusiness/DataProvider/Server.swift [hunk 5]
```diff
@@ -7,29 +7,102 @@
...
+            return .init(url: url)
         }
+        components.queryItems = queryItems + [
+            URLQueryItem(name: "onlyChapters", value: "true"),
+            URLQueryItem(name: "vector", value: vector)
+        ]
+        return .init(url: components.url ?? url)
+    }
+
+    private static func standardResizedImageUrl(_ url: URL, width: ImageWidth, baseUrl: URL, queryItems: [URLQueryItem]) -> URL {
```
Comment (waliid) [inline Sources/CoreBusiness/DataProvider/Server.swift:81]:
```suggestion
    private static func resizedImageUrl(_ url: URL, width: ImageWidth, baseUrl: URL, queryItems: [URLQueryItem]) -> URL {
```

File: Sources/CoreBusiness/DataProvider/Server.swift (hunk 4 above)
Comment (defagos) [reply Sources/CoreBusiness/DataProvider/Server.swift:69]: Applied, thanks.

File: Sources/CoreBusiness/DataProvider/Server.swift (hunk 5 above)
Comment (defagos) [reply Sources/CoreBusiness/DataProvider/Server.swift:81]: Applied, thanks.

File: Sources/CoreBusiness/DataProvider/Server.swift (hunk 3 above)
Comment (defagos) [reply Sources/CoreBusiness/DataProvider/Server.swift:61]: Would be misleading IMHO. URN is just an identifier for content on a server, not the server itself. Maybe `custom` is not the proper word for it but we need a way to define custom servers which behave like an IntegrationLayer (e.g. MMF or SAM).

Comment (waliid) [reply Sources/CoreBusiness/DataProvider/Server.swift:61]: Should we rename the urn closure parameter by identifier then?

File: Demo/Sources/Model/ServerSetting.swift (hunk 1 above)
Comment (defagos) [reply Demo/Sources/Model/ServerSetting.swift:26]: Done.

File: Sources/CoreBusiness/DataProvider/Server.swift (hunk 2 above)
Comment (defagos) [reply Sources/CoreBusiness/DataProvider/Server.swift]: I removed APIs for free construction with closures. Was overkill.

File: Sources/CoreBusiness/DataProvider/Server.swift (hunk 3 above)
Comment (defagos) [reply Sources/CoreBusiness/DataProvider/Server.swift]: I exposed the public initializer instead.

**Comment by defagos** [comment]
I revisited my original implementation which was providing closures to make it possible to build media composition requests and custom image resizing to make it simpler. In practice having a custom service posing as the IL is only useful internally anyway and requires deep understanding of endpoints and metadata formats.

We currently have SAM and MMF posing as the IL, as well as specific ways to simulate a user abroad. Those are implemented in the demo (and thus not provided in the Core Business public API). When SAM is released this mechanism will not be useful anymore anyway, so I guess it is better to keep things simple.

**Review by waliid — APPROVED** [review]

**Pull request merged**